HIP/hip-904-frictionless-airdrop.md

```diff
@@ -0,0 +1,711 @@
+
+The record file emitted from a consensus node will contain
+
+`tokenTransferLists` containing 
+
+- Alice debit of 5 T1 tokens → Bob  credit of 5 T1 tokens
+- Alice debit of NFT serials 1 & 2 of token T2 → Bob credit of NFT serials 1 & 2 of token T2
+- Alice debit of 10 T3 tokens → Carol  credit of 10 T3 tokens
+- Alice debit of NFT serials 3 & 4 of token T4 → Carol credit of NFT serials 1 & 2 of token T4
+
+`automatic_token_associations` containing
+
+- Carols automatic association with token T3
+- Carols automatic association with token T4
+
+`pending_airdrops` containing
+
+- Alices intent to transfer 15 T5 tokens to Dave
+- Alices intent to transfer NFT serials 5 & 6 of token T6 to Dave
+
+Note: In the transaction body Alice had split her intended transfers to Dave in two but in the result exposed by the record they were combined. TokenAirdrops may be combined in an additive fashion. It is not possible to subtract. For this a sender may cancel and reissue their airdrop 
+
+Scenario 2: If Dave submits a TokenClaim for token T5 and a Token claim of serial 5 of T6
+
+`tokenTransferLists`
+
+- Alice debit of 15 T5 tokens → Dave credit of 15 T5 tokens
+- Alice debit of NFT serial 5 of token T6 → Dave credit of NFT serial 5 of token T6
+
+`automatic_token_associations` containing
+
+- Daves automatic association with token T5
+- Dave automatic association with token T6
+
+`pending_airdrops` will be empty
+
+Note: NFT serial 6 of T6 is left in pending airdrop, but all of the pending fungible amount of T5 was transferred 
+
+Scenario 3: Eric submits a TokenCancelAirdrop for token T7 and a TokenCancelAirdrop of serial 7 of T8
+
+`tokenTransferLists` will be empty
+
+`automatic_token_associations` will be empty
+
+`pending_airdrops` will be empty, details are in transaction body
+
+Note: NFT serial 6 of T6 is left in pending airdrop, but all of the pending fungible amount of T5 was transferred 
+
+#### Reject
+
+```protobuf
+
+/**
+ * Reject undesired tokens. Support up to a maximum of 10 trasnfers to treasurers
+ */
+message TokenRejectTransactionBody {
+	repeated TokenID fungible_token_ids = 1;
+
+	repeated NftID nft_ids = 2;
+}
+```
+
+#### Token Service
+
+```protobuf
+service TokenService {
+
+    ...
+
+    /**
+    * Airdrops one or more tokens
+    */
+    rpc airdropTokens (Transaction) returns (TransactionResponse);
+
+    /**
+    * Cancels one or more pending airdrops
+    */
+    rpc cancelAirdrop (Transaction) returns (TransactionResponse);
+
+    /**
+    * Claims a pending airdrop. This transaction must be signed by the account
+    * receiving the airdrop (in essence, airdrops on accounts without open
+    * auto-association slots behave as though receiver-sig-required is true for
+    * airdrops).
+    */
+    rpc claimAirdrop (Transaction) returns (TransactionResponse);
+
+    /**
+    * Reject tokens in balance
+    */
+    rpc rejectToken (Transaction) returns (TransactionResponse);
```

Comment:
suggest `rejectAirdrop` or `rejectTokenAirdrop`

Reply:
As mentioned above, wouldn't it be a good idea for a receiver to be able to reject also an airdrop in pending state? Even if the `tokenReject` remains to be an operation only for tokens in a given balance, we can introduce a separate `tokenAirdropReject` function for rejecting only airdrops.